File: src/nl/rubensten/texifyidea/inspections/latex/LatexOverInsteadOfFracInspection.java
package nl.rubensten.texifyidea.inspections.latex;

import com.intellij.codeInspection.InspectionManager;
import com.intellij.codeInspection.LocalQuickFix;
import com.intellij.codeInspection.ProblemDescriptor;
import com.intellij.codeInspection.ProblemHighlightType;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiDocumentManager;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import nl.rubensten.texifyidea.index.LatexCommandsIndex;
import nl.rubensten.texifyidea.insight.InsightGroup;
import nl.rubensten.texifyidea.inspections.TexifyInspectionBase;
import nl.rubensten.texifyidea.lang.magic.MagicCommentScope;
import nl.rubensten.texifyidea.psi.LatexCommands;
import nl.rubensten.texifyidea.psi.LatexMathContent;
import nl.rubensten.texifyidea.psi.LatexPsiUtil;
import nl.rubensten.texifyidea.util.DocumentsKt;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.EnumSet;
import java.util.List;
import java.util.Set;

/**
 * @author Ruben Schellekens
 */
public class LatexOverInsteadOfFracInspection extends TexifyInspectionBase {

    @NotNull
    @Override
    public InsightGroup getInspectionGroup() {
        return InsightGroup.LATEX;
    }

    @Nls
    @NotNull
    @Override
    public String getDisplayName() {
        return "Discouraged use of \\over";
    }

    @NotNull
    @Override
    public String getInspectionId() {
        return "OverInsteadOfFrac";
    }

    @NotNull
    @Override
    public Set<MagicCommentScope> getOuterSuppressionScopes() {
        return EnumSet.of(MagicCommentScope.COMMAND);
    }

    @NotNull
    @Override
    public List<ProblemDescriptor> inspectFile(@NotNull PsiFile file, @NotNull InspectionManager
            manager, boolean isOntheFly) {
        List<ProblemDescriptor> descriptors = descriptorList();

        Collection<LatexCommands> commands = LatexCommandsIndex.Companion.getItems(file);
        for (LatexCommands command : commands) {
            if ("\\over".equals(command.getName())) {
                descriptors.add(manager.createProblemDescriptor(
                        command,
                        "Use of \\over is discouraged",
                        new OverToFracFix(),
                        ProblemHighlightType.GENERIC_ERROR_OR_WARNING,
                        isOntheFly
                ));
            }
        }

        return descriptors;
    }

    private class OverToFracFix implements LocalQuickFix {

        @Nls
        @NotNull
        @Override
        public String getFamilyName() {
            return "Convert to \\frac";
        }

        @Override
        public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor descriptor) {
            PsiElement element = descriptor.getPsiElement();
            if (!(element instanceof LatexCommands)) {
                return;
            }

            // Find elements to put in numerator and denominator.
            LatexCommands cmd = (LatexCommands)element;
            PsiElement content = cmd.getParent().getParent();
            if (content instanceof LatexMathContent) {
                content = cmd.getParent();
            }
            PsiElement previous = LatexPsiUtil.getPreviousSiblingIgnoreWhitespace(content);
            PsiElement next = LatexPsiUtil.getNextSiblingIgnoreWhitespace(content);

            String before = (previous == null ? "" : previous.getText());
            String after = (next == null ? "" : next.getText());
            String replacement = String.format("\\frac{%s}{%s}", before, after);

            Document document = PsiDocumentManager.getInstance(project).getDocument(element.getContainingFile());

            // Delete denominator.
            if (next != null) {
                DocumentsKt.deleteElement(document, next);
            }

            // Replace command.
            TextRange range = ((LatexCommands)element).getCommandToken().getTextRange();
            if (document != null) {
                document.replaceString(range.getStartOffset(), range.getEndOffset(), replacement);
            }

            // Replace numerator.
            if (previous != null) {
                DocumentsKt.deleteElement(document, previous);
            }
        }
    }
}

File: src/nl/hannahsten/texifyidea/inspections/latex/LatexOverInsteadOfFracInspection.kt
package nl.hannahsten.texifyidea.inspections.latex

import com.intellij.codeInspection.InspectionManager
import com.intellij.codeInspection.LocalQuickFix
import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiDocumentManager
import com.intellij.psi.PsiFile
import nl.hannahsten.texifyidea.index.LatexCommandsIndex
import nl.hannahsten.texifyidea.insight.InsightGroup
import nl.hannahsten.texifyidea.inspections.TexifyInspectionBase
import nl.hannahsten.texifyidea.lang.magic.MagicCommentScope
import nl.hannahsten.texifyidea.psi.LatexCommands
import nl.hannahsten.texifyidea.psi.LatexMathContent
import nl.hannahsten.texifyidea.psi.LatexPsiUtil.getNextSiblingIgnoreWhitespace
import nl.hannahsten.texifyidea.psi.LatexPsiUtil.getPreviousSiblingIgnoreWhitespace
import nl.hannahsten.texifyidea.util.deleteElement
import org.jetbrains.annotations.Nls
import java.util.EnumSet

/**
 * @author Hannah Schellekens
 */
class LatexOverInsteadOfFracInspection : TexifyInspectionBase() {
    override val inspectionGroup: InsightGroup
        get() = InsightGroup.LATEX

    @Nls
    override fun getDisplayName(): String {
        return "Discouraged use of \\over"
    }

    override val inspectionId: String
        get() = "OverInsteadOfFrac"

    override val outerSuppressionScopes: Set<MagicCommentScope>
        get() = EnumSet.of(MagicCommentScope.COMMAND)

    override fun inspectFile(
        file: PsiFile,
        manager: InspectionManager,
        isOntheFly: Boolean
    ): List<ProblemDescriptor> {
        val descriptors = descriptorList()
        val commands = LatexCommandsIndex.getItems(file)
        for (command in commands) {
            if ("\\over" == command.name) {
                descriptors.add(
                    manager.createProblemDescriptor(
                        command,
                        "Use of \\over is discouraged",
                        OverToFracFix(),
                        ProblemHighlightType.GENERIC_ERROR_OR_WARNING,
                        isOntheFly
                    )
                )
            }
        }
        return descriptors
    }

    private inner class OverToFracFix : LocalQuickFix {
        @Nls
        override fun getFamilyName(): String {
            return "Convert to \\frac"
        }

        override fun applyFix(
            project: Project,
            descriptor: ProblemDescriptor
        ) {
            val element = descriptor.psiElement as? LatexCommands ?: return

            // Find elements to put in numerator and denominator.
            var content = element.parent.parent
            if (content is LatexMathContent) {
                content = element.parent
            }
            val previous = getPreviousSiblingIgnoreWhitespace(content!!)
            val next = getNextSiblingIgnoreWhitespace(content)
            val before = if (previous == null) "" else previous.text
            val after = if (next == null) "" else next.text
            val replacement = String.format("\\frac{%s}{%s}", before, after)
            val document =
                PsiDocumentManager.getInstance(project).getDocument(element.containingFile)

            // Delete denominator.
            if (next != null) {
                document!!.deleteElement(next)
            }

            // Replace command.
            val range = element.commandToken.textRange
            document?.replaceString(range.startOffset, range.endOffset, replacement)

            // Replace numerator.
            if (previous != null) {
                document!!.deleteElement(previous)
            }
        }
    }
}